markdown/dt-otel/dt-otel.md

```diff
@@ -779,6 +808,176 @@ The second span was generated by the PubSub+ Broker when the message was receive
+            //Set more attributes as needed
+            .setAttribute("myReceiveKey", "receiveValue1")
+            //.setAttribute(...)
+            .setParent(extractedContext).startSpan();
+
+    //... and then we do processing and have another span
+    try {
+        final Span processSpan = JcsmpTracingUtil.tracer.spanBuilder(SERVICE_NAME + " " + SpanAttributes.MessagingOperation.PROCESS).setSpanKind(SpanKind.CLIENT)
+                //Set more attributes as needed
+                .setAttribute("myProcessingKey", "postProcessInformation")
+                //.setAttribute(...)
+                .setParent(Context.current().with(receiveSpan)) // make RECEIVE span be parent
+                .startSpan();
+// continued below...
+```
+After creating both the receive and process spans, completing our custom processing, and ACK'ing the message we can go ahead and end our process and receive spans (order is important here).
+Only after ending the current span is the data exported to the configured Opentelemetry collector.
+<aside>
+The OpenTelemetry collector is brought in to the application via the _opentelemetry-exporter-otlp_ library. Default settings are used to export
+to the collector executing in Docker.
+</aside>
+
+```java
+//...continued from above
+        try {
+            msgRecvCounter++;
+            if (message.getRedelivered()) {  // useful check
+                // this is the broker telling the consumer that this message has been sent and not ACKed before.
+                 // this can happen if an exception is thrown, or the broker restarts, or the network disconnects
+                 // perhaps an error in processing? Should do extra checks to avoid duplicate processing
+                 hasDetectedRedelivery = true;
+            }
+             // Messages are removed from the broker queue when the ACK is received.
+             // Therefore, DO NOT ACK until all processing/storing of this message is complete.
+             // NOTE that messages can be acknowledged from a different thread.
+             message.ackMessage();  // ACKs are asynchronous
+        } catch (Exception e) {
+         processSpan.recordException(e); //Span can record exception if any
+         processSpan.setStatus(StatusCode.ERROR, e.getMessage()); //Set span status as ERROR/FAILED
+        } finally {
+         processSpan.end(); //End processSpan. Span data is exported when span.end() is called.
+        }
+    } finally {
+     receiveSpan.end(); //End receiveSpan. Span data is exported when span.end() is called.
+    }
+}
+```
+
+We've provided a jar containing the manual implementation code from above, to run the jar and subscribe to the `q` queue on the broker, execute the command below from the tracing-codelab directory.
+```bash
+java -jar solace-samples-jcsmp-subscriber-manual-instrumentation-1.0-SNAPSHOT.jar tcp://localhost:55557 default default default
```

Comment:
File is currently available in the target directory [here](https://solacesystems-my.sharepoint.com/personal/hank_spencer_solace_com/_layouts/15/onedrive.aspx?FolderCTID=0x0120000F184F3829FEEA4782B84E73C4226B07&id=%2Fpersonal%2Fhank%5Fspencer%5Fsolace%5Fcom%2FDocuments%2FDT%20OTEL%20Code%20Lab%2Fmanual%2Dinstrumentation%2Dsubscriber&view=0). We'll need to copy the final version of both the unpackaged code and the jar to the tracing-codelab zip that is currently downloaded at the start of the codelab.

Unpackaged code can be ran as a java class. I'm assuming we'll want to provide the users with a .jar file so they don't have to package anything up.